rfd/0078-login-rules.md

```diff
@@ -0,0 +1,703 @@
+
+### When login rules will be parsed and evaluated
+
+Login rules will be parsed and evaluated during each user login.
+This will occur after the SSO provider has returned its assertions/claims, and
+before Teleport maps these to internal Teleport roles via `attributes_to_roles` or
+`claims_to_roles` so that transformed traits can be used for this mapping.
+
+An eventual extension to login rules could also return the desired set of
+Teleport roles which the user should have.
+
+During login, the auth server will load all `login_rule` resources in the
+cluster, sort them by `priority` and `name`, and apply all of them in order.
+
+### Trusted clusters
+
+Since login rules will be evaluated during login and the resulting traits will
+be embedded in the user's certificates, they only need to be created in the root
+cluster and the transformed traits will be visible and usable in all leaf
+clusters.
+
+### Creating and modifying login rules
+
+Login rules will be a new backend resource `login_rule`.
+
+They should be written in a YAML file and created by the usual means of
+`tctl create resource.yaml`. `tctl get login_rule/example` can be used to
+fetch the current resource, and `tctl rm login_rule/example` can be used
+to delete the resource.
+
+### Validating and testing login rules
+
+A new `tctl` command will be introduced, which can be used to experiment with
+login rules and test what their effect will be before deploying them to
+your cluster.
+
+```bash
+$ tctl test login_rule \
+  --load login_rule1.yaml \
+  --load login_rule2.yaml \
+  --load-from-cluster \
+  --input_traits '{"groups": ["splunk"], "email": "nic@goteleport.com", "username": "nklaassen"}'
+```
+
+You can load multiple yaml files with `--load resource.yaml`, optionally load
+existing trait transforms from the cluster with `--load-from-cluster`, and
+provide a set of input traits to test with.
+
+The command will report any syntax errors, and will print the output traits for
+the given input.
+
+### Protobuf Definitions
+
+The `login_rule` resource and associated CRUD RPCs will be added to a new
+package `teleport/loginrule/v1`.
+
+```
+// loginrule.proto
+...
+
+// LoginRule is a resource to configure rules and logic which should run during
+// Teleport user login.
+message LoginRule {
+  // kind is a resource kind
+  string kind = 1;
+  // sub_kind is an optional resource sub kind, used in some resources
+  string sub_kind = 2;
+  // version is a resource version
+  string version = 3;
+  // metadata is resource metadata
+  types.Metadata metadata = 4;
+  // spec is the login rule specification
+  LoginRuleSpec spec = 5;
+}
+
+// LoginRuleSpec is a login rule specification.
+message LoginRuleSpec {
+  // priority is the priority of the login rule relative to other login rules
+  // in the same cluster. Login rules with a lower numbered priority will be
+  // evaluated first.
+  string priority = 1;
```

Comment:
Did you pick string to avoid the zero value of zero? I'm concerned a string field allows for many more invalid inputs than an integer.

Reply:
no that's a mistake/typo, it should be some integer type, I'll fix